### ISSUE TITLE
Bot (suddenly) stops listening, even with example script

### ISSUE THREAD

**Issue by funiel** [imayhaveborkedit/discord-ext-voice-recv#27]
Hi! Thanks for your hard work on the sinks. Been using this extension for a while and the callback suddenly stopped working a couple weeks ago. So I tried running the provided basic example. Still, the callback stops receiving packets after a couple of seconds.

## Steps to reproduce
* Installed Python Ver: 3.12.7
* Installed `discord-py`, `discord-py[voice]`, `discord-ext-voice-recv`
* Update `examples/recv.py` so the commands get recognized by discord.py
* Run bot
* Bot joins voice channel
* Runs for a couple seconds, then stops receiving packets, continues to stay in the channel

**Comment by funiel** [comment]
Okay so this might be related to deprecated encryption modes since the dates do somewhat coincide? Not entirely sure.

https://github.com/Rapptz/discord.py/pull/9953

**Comment by OCT-Burnz** [comment]
Can confirm this behavior, been debugging this issue as well. The callback isn't triggered anymore after a while.

**Comment by Pl8tinium** [comment]
anyone found a fix yet?

**Comment by Pl8tinium** [comment]
it seems like the socket connection has an issue, maybe it timeouts after a while. The callback of the socket is not triggered anymore after a couple of seconds

![image](https://github.com/user-attachments/assets/bb44279a-5168-440b-a10c-2ec65ade5443)

**Comment by Pl8tinium** [comment]
mmh it is probably an issue with the socket connection, but not the connection itself, atleast the VoiceConnectionState reports back that its still connected 

![image](https://github.com/user-attachments/assets/fdfcdbc5-d32b-418c-9983-bf22b49fee26)

idk, maybe i'll give this a shot another day..

**Comment by Naivy-dev** [comment]
Guys i'm having the same problem, still didn't find the fix, but for some reason, i have two friends that doesn't have this problem. We tried a bunch of different things and still didn't found. We're using Python 3.10.16

**Comment by Pl8tinium** [comment]
i have had it working for a brief period as well, i just revisited my script a couple months later and dont remember which python version/ packages ive been using. I do know that my commit history says that ive been using discord.py 1.7.3, but i wasnt able to reproduce my prior environment (which i do remember had a lot of different package installation quirks like mismatching versions and using the -no-deps flag for installing packages...)

so in case anyone gets it working with discord.py version 1.7.3, theres a good chance it'll work

**Comment by brianle-riceroll** [comment]
#8 

I have found a workaround solution that is found in the above issue

**Comment by Beregoz** [comment]
> [#8](https://github.com/imayhaveborkedit/discord-ext-voice-recv/issues/8)
> 
> I have found a workaround solution that is found in the above issue.

This did not work for me.

I can see the bot is able notice if im speaking in the channel, but it is not converting the speech to text. "Bad speech chunk".

**Comment by brianle-riceroll** [comment]
> > [#8](https://github.com/imayhaveborkedit/discord-ext-voice-recv/issues/8)
> > I have found a workaround solution that is found in the above issue.
> 
> This did not work for me.
> 
> I can see the bot is able notice if im speaking in the channel, but it is not converting the speech to text. "Bad speech chunk".

I can't tell if you're having issues with the speech-to-text or if it's the audio sinks you used. Which audio sinks are you using and what speech-to-text solution are you using?

You can check if the audio sinks are actually receiving audio by writing a simple print statement in the write() method. If the bot is actually listening to you and receiving audio packets from you, it will spit a crap ton of print statements. Otherwise, it will not spit out a bunch of print statements:

![Image](https://github.com/user-attachments/assets/2983b643-2d55-4474-85bd-f0b2f21ec6d3)
_Example using BasicSink_

When testing this, make sure you test it over a decent period of time. When I tested, it took around 10-15 minutes until it suddenly stopped receiving audio packets.

I also haven't encountered the "Bad Speech Chunk" error before. I use [openai/whisper](https://github.com/openai/whisper) as my speech-to-text solution and never encountered that.

**Comment by Beregoz** [comment]
> > > [#8](https://github.com/imayhaveborkedit/discord-ext-voice-recv/issues/8)
> > > I have found a workaround solution that is found in the above issue.
> > 
> > 
> > This did not work for me.
> > I can see the bot is able notice if im speaking in the channel, but it is not converting the speech to text. "Bad speech chunk".
> 
> I can't tell if you're having issues with the speech-to-text or if it's the audio sinks you used. Which audio sinks are you using and what speech-to-text solution are you using?
> 
> You can check if the audio sinks are actually receiving audio by writing a simple print statement in the write() method. If the bot is actually listening to you and receiving audio packets from you, it will spit a crap ton of print statements. Otherwise, it will not spit out a bunch of print statements:
> 
> ![Image](https://github.com/user-attachments/assets/2983b643-2d55-4474-85bd-f0b2f21ec6d3) _Example using BasicSink_
> 
> When testing this, make sure you test it over a decent period of time. When I tested, it took around 10-15 minutes until it suddenly stopped receiving audio packets.
> 
> I also haven't encountered the "Bad Speech Chunk" error before. I use [openai/whisper](https://github.com/openai/whisper) as my speech-to-text solution and never encountered that.

I'm using the  SpeechRecognitionSink with azure.
` vc.listen(voice_recv.extras.SpeechRecognitionSink(process_cb=process_audio, default_recognizer="azure", text_cb=got_text))`

In my tests the bot stops working after 10-15 minutes and it only drops random user at a time. So in a channel with 4 users, it will stop listening one-by-one.

**Comment by brianle-riceroll** [comment]
> > > > [#8](https://github.com/imayhaveborkedit/discord-ext-voice-recv/issues/8)
> > > > I have found a workaround solution that is found in the above issue.
> > > 
> > > 
> > > This did not work for me.
> > > I can see the bot is able notice if im speaking in the channel, but it is not converting the speech to text. "Bad speech chunk".
> > 
> > 
> > I can't tell if you're having issues with the speech-to-text or if it's the audio sinks you used. Which audio sinks are you using and what speech-to-text solution are you using?
> > You can check if the audio sinks are actually receiving audio by writing a simple print statement in the write() method. If the bot is actually listening to you and receiving audio packets from you, it will spit a crap ton of print statements. Otherwise, it will not spit out a bunch of print statements:
> > ![Image](https://github.com/user-attachments/assets/2983b643-2d55-4474-85bd-f0b2f21ec6d3) _Example using BasicSink_
> > When testing this, make sure you test it over a decent period of time. When I tested, it took around 10-15 minutes until it suddenly stopped receiving audio packets.
> > I also haven't encountered the "Bad Speech Chunk" error before. I use [openai/whisper](https://github.com/openai/whisper) as my speech-to-text solution and never encountered that.
> 
> I'm using the SpeechRecognitionSink with azure. ` vc.listen(voice_recv.extras.SpeechRecognitionSink(process_cb=process_audio, default_recognizer="azure", text_cb=got_text))`
> 
> In my tests the bot stops working after 10-15 minutes and it only drops random user at a time. So in a channel with 4 users, it will stop listening one-by-one.

Yeah, I never touched that audio sink. Can you implement a command that sends an empty packet like the one shown below?

![Image](https://github.com/user-attachments/assets/50c5ed6e-1434-490a-9f7b-7b50f86a29af)

Once your bot stops working, execute that command in Discord to see if it works. This was the command I created that initially fixed my listening issues.

**Comment by Beregoz** [comment]
I did have an implementation which would do that periodically and that didnt help.

**Comment by brianle-riceroll** [comment]
@Beregoz 
I'm going to use @SilverBeamx idea to see if that solves your issue. Go to this section of the code and add a print statement to detect a malformed packet that your bot might have picked up:
![Image](https://github.com/user-attachments/assets/243f1215-6551-4d90-95db-2a6c926ce098)

Then in the same file, create a class like this to handle the malformed packet:
![Image](https://github.com/user-attachments/assets/732ac873-ec96-4da3-9a96-4e358f36939e)

And at the very bottom, map 0 to the FakeRTCPPacket class:
![Image](https://github.com/user-attachments/assets/50bfe8f9-2c2c-429c-940c-de67cd9fbd45)

I'm still currently testing @SilverBeamx solution as I did encounter a rare moment where the bot stopped receiving my voice in the middle of recording (but completely fine if there are long moments of silence w/ the silent packet solution). I'm hoping this will solve your issue.

**Comment by brianle-riceroll** [comment]
Just an update. After hours of debugging, my findings show that the issue may be somewhere in the opus.py and buffer.py module. This will be a very long technical analysis and report. It seems that this issue has two parts:
1. Listening Sockets closing if the bot does not send a packet (fixed with my workaround).
2. The bot does listen and receive audio packets but does not process them after ~10-15min (The current issue).

The callback in the AudioReader class (in the reader.py module) actually continuously receives packets, even after the discord bot "stops listening" to us, so the bot is receiving packets, it's just not processing them because there is an issue with the decoder. We can prove this as once the bot stops receiving audio from us if the **user** reconnects, then the bot once again receives and processes audio again. When a new user connects to a voice channel or reconnects, a new "ssrc" value is assigned to each user and these ssrc values are unique every time. So this definitely is not an issue with callback functions.


I think the issue lies somewhere in the opus.py module. @Beregoz mentioned:
 "**In my tests the bot stops working after 10-15 minutes and it only drops a random user at a time. So in a channel with 4 users, it will stop listening one-by-one.**" 

This is because each user is assigned a new decoder thread (specifically named "**decoder-ssrc-{ssrc number}**"). And if one decoder thread has an issue with processing packets, then the user that is assigned to that thread will no longer have their packets processed.
![Image](https://github.com/user-attachments/assets/c8ce81d8-6c68-41b5-8b51-0e1d33d28fbe)
This image has one ssrc decoder thread (that is assigned to me, where I am the only one in the voice channel)

When I was debugging and testing, I added breakpoints in opus.py. I specifically added a breakpoint in the _get_next_packet() method of the PacketDecoder class:
![Image](https://github.com/user-attachments/assets/2141675c-4a00-4f7a-bcd7-8d3c580937eb)

Basically, after 10-15min, the bot will stop processing our audio packets and once I reach that point, the packet (which calls self._buffer.pop() to get its value) is set to None. Normally, this would not be the case and the packet variable would not be None, allowing the thread to call **self._process_packet()** in the **_do_run()** method:
![Image](https://github.com/user-attachments/assets/712c757c-fe1b-47f0-b62e-af5c7ee9e7dc)

We want the **_process_packet(**) method to be called because, inside that method, it calls our AudioSink's **write()** method:
![Image](https://github.com/user-attachments/assets/9e59a32a-4bd7-4ef6-aead-de412a1551d4)
However, this will never get called if our **packet** variable in **_do_run()** method is None.

Looking into the pop method implementation, we have a variable called "**ok**", which calls the **self._has_item.wait()** method. The self._has_item is a threading.Event() class, which is basically like a Mutex or thread lock. These are set to lock (False) and unlock (True) by the set() and clear() methods, which are called throughout the buffer.py module. As you can probably guess, the **ok** variable is None, which is passed back into **packet**:
![Image](https://github.com/user-attachments/assets/181bfc22-5c9c-484a-8b99-0613476595ba)

I still need to do further debugging and testing, but my prediction is that either the locks are not being locked and unlocked properly or there is some kind of issue with the HeapJitterBuffer class. @Beregoz I hope this helps you a bit in debugging and I'll spend some time trying to find the problem and coming up with fixes for it.

I know they've been inactive for a while but... @imayhaveborkedit, are you aware of this issue so far?

**Comment by brianle-riceroll** [comment]
> Unfortunately this did not fix my problem. I need to debug it with better focus. My current workaround is to reconnect the voice bot when i receive an error.

@Beregoz I got a question. I may have found a fix but I discovered something interesting. Did you install discord-ext-voice-recv via 
`python -m pip install discord-ext-voice-recv` or `python -m pip install git+https://github.com/imayhaveborkedit/discord-ext-voice-recv`?

If you did `python -m pip install discord-ext-voice-recv`, try uninstalling it and install it with `python -m pip install git+https://github.com/imayhaveborkedit/discord-ext-voice-recv` (backup the voice_recv folder if you have to). Apparently, the former installation method does not install the latest version of this extension. The latter installation installs the latest version of this extension. Give that a try and test out your bot.

**Comment by madman37287** [comment]
I'm deep in this rabbit hole too.  I thought my issue was different, but now that I have done some source diving, I think our issues are the same.  Unfortunately, the latest git doesnt cut it for me.  It seems something in the refactors added enough overhead to slow the user recordings to a snail's pace, where the version contained in PyPI is near instantaneous

**Comment by brianle-riceroll** [comment]
> I'm deep in this rabbit hole too. I thought my issue was different, but now that I have done some source diving, I think our issues are the same. Unfortunately, the latest git doesnt cut it for me. It seems something in the refactors added enough overhead to slow the user recordings to a snail's pace, where the version contained in PyPI is near instantaneous

Yeah, the latest git version is slow as a snail's pace which is something I don't want for my bot. The fix I have works well for 1 person in a call with the bot, but gets deadlocks if more users join. I decided to give up and port my discord bot from discord.py to pycord due to this issue. 

Pycord doesn't seem to have any issues with receiving audio (except the listening socket closing but this is fixed with my workaround I posted here). Their audio sinks classes are very similar as well so it was very easy for me to port everything over.

**Comment by madman37287** [comment]
Oh really?  I was under the impression that Pycord only had streaming functionality, not recording.  I'll check that out right now, that sounds like exactly what I have been looking for.

**Comment by madman37287** [comment]
Looking at Pycord now, does it allow you to do real-time processing of the data?  The guides only seem to show saving as files.

**Comment by brianle-riceroll** [comment]
> Looking at Pycord now, does it allow you to do real-time processing of the data? The guides only seem to show saving as files.

I basically have my own custom sink class (that derives from [Pycord's Sink class](https://github.com/Pycord-Development/pycord/blob/master/discord/sinks/core.py)). My custom sink class basically is able to get audio from each user and store them in a dictionary so that I can process each audio from each user separately later (My bot can discern who said what). I'm storing the audio data in bytes so that I can write them into a wav form later to be transcribed via a callback function.

Just make a class that derives from it and implement the methods in Pycord's Sink class. Here is an example of how you can make a custom sink that stores audio data of each user:
![Image](https://github.com/user-attachments/assets/8e714863-c60a-4d96-8f56-3ae2184c19f9)

**Comment by madman37287** [comment]
Nice, thats pretty much exactly what i am doing with the packets once i get them anyways.  Also this morning i found the root of the issue.  So if you are curious or anybody else is beating their head against the table, here is the root cause:

TL;DR: 
-----------------------------------------
There is a 16bit limitation to the sequence.  Once we surpass 65535, we go back to 0 and start over.  The code TRIES to account for this in JitterBuffer.push() by incrementing self._generation += 1 to add a multiplier to keep the counting accurate.  However, due to the sequence of events, JitterBuffer.pop() will set self._last_tx = 65535 + 65536.  The cleanup function will then remove everything that push() tries to enter into the buffer.
-----------------------------------------

TLDR FIX:
-----------------------------------------
file: buffer.py

in HeapJitterBuffer.pop():

```
    def pop(self, *, timeout=1.0):
        """
        If timeout is a positive number, wait as long as timeout for a packet
        to be ready and return that packet, otherwise return None.
        """

        ok = self._has_item.wait(timeout)
        if not ok:
            print("not ok")
            return None

        if self._prefill > 0:
            return None

        # This function should actually be redundant but i'll leave it for now
        packet = self._pop_if_ready()

        if packet is not None:
            seq = self._get_seq(packet)
            print(f"pop test: {self._last_tx} | {seq}")

            ##################
            # REMOVE
            ##################
            # self._last_tx = self._get_seq(packet)
            ##################


            ##################
            # ADD
            ##################
            # Detect large jump in data sequence caused by self._generation += 1
            # Tried to 32768 is chosen to mirror push function. Any large values up to 65536 should work.
            if seq + 32768 < self._last_tx:
                self._last_tx = self._get_seq(packet) - 65536
            else:
                self._last_tx = self._get_seq(packet)
            ##################

        self._update_has_item("pop")
        return packet
```
-----------------------------------------

-----------------------------------------
Full Explanation
-----------------------------------------
I copied the library and renamed it "debug" so i could make edits as I pleased.  I then created a test script to exercise the "push" and "pop" functions of the jitter buffer.  To be honest, I was confident that the issue was a result of improper thread handling, since "push" and "pop" both modify the buffer in different threads, and that would explain the "randomness" i was seeing.  I planned on modifying the test script to do push and pop in multithreading, but it turns out I didn't have to.  here is the code that replicates the issue:

```
# Note: with fix, this file fails after sequence 131,070 (65535*2).  
# This is because my test script is re-using the same timestamp in the packet over and over.  
# This is an issue with the test script, not the solution.

import discord.ext.voice_recv_debug as vc
from discord.ext.voice_recv_debug.rtp import RTPPacket as rtp

class BufferTest():
    def __init__(self):
        self.buffer = vc.buffer.HeapJitterBuffer()
        self.curval  = b'\x90x\x13\x10' # this guy incrememnts in each packet.  Checking for packet increment is used in JitterBuffer push/pop
        self.package_content = b'\xca\xe2\x17\xcc\x00\x017\xfa\xba\xec"|\xa2\xf4\x88^G\x95\x0e\\\xbd};\xe0\x92\xeb\xb4\xd8OW\x1b\x7f\xe6:c\xa5\n\xea\xc6\x08\x00\x00\x00' # copied from real packet

        self.count = 0 # can be printed to somewhat see issue without making modifications to library
    
    def getRTP(self):
        packet = rtp(self.curval + self.package_content)
        self.curval = int(int.from_bytes(self.curval, "big") + 1).to_bytes(4, "big")
        self.count += 1
        return packet
    
    def push(self):
        packet = self.getRTP()
        return self.buffer.push(packet)
        
    def pop(self):
        return self.buffer.pop()


if __name__ == "__main__":
    buffer = BufferTest()

    while True:
        buffer.push()
        buffer.pop()

        # print(buffer.count)

```

If you print the sequence value in push(), you will see that we fail everytime we cross the 16 bit threshold value of 65535.  This is because of this code here:

```
    def _get_seq(self, packet: RTPPacket) -> int:
        return packet.sequence + 65536 * self._generation

    def push(self, packet: RTPPacket) -> bool:
        """
        Push a packet into the buffer.  If the packet would make the buffer
        exceed its maxsize, the oldest packet will be dropped.
        """

        seq = self._get_seq(packet)

        # if the seq has rolled over, it'll be ~65535 lower than the generation ts
        if seq + 32768 < self._last_rx and packet.timestamp > self._generation_ts:
            self._generation += 1
            self._generation_ts = packet.timestamp
            seq = self._get_seq(packet)
```
When we hit 65535, self._generation is incremented by a value of 1.  When this happens, _get_seq will add 65536 for every time we loop.  However, the way this is implemented breaks something else along the way.  We have two member values called "self._last_rx" and "self._last_tx" which check to ensure packets are following their given order of operations.  in this situation, last_rx is being updated properly.  last_tx  is not.  last_tx is handled within the pop function shown here:

```
    def pop(self, *, timeout=1.0):
        """
        If timeout is a positive number, wait as long as timeout for a packet
        to be ready and return that packet, otherwise return None.
        """

        ok = self._has_item.wait(timeout)
        if not ok:
            print("not ok")
            return None

        if self._prefill > 0:
            return None

        # This function should actually be redundant but i'll leave it for now
        packet = self._pop_if_ready()

        if packet is not None:
            seq = self._get_seq(packet)
            self._last_tx = self._get_seq(packet)

        self._update_has_item("pop")
        return packet
```
When we pop the packet associated with sequence 65535, packet 65536 has already been pushed (there is always at least 1 packet in the buffer).  This is where the problem occurs.  In the line `self._last_tx = self._get_seq(packet)`, last_tx will be set to 65535 + 65536.  When push goes to shove in its next packet, this incorrect last_tx value will cause the cleanup function to remove all packets with a lower sequence than the sequence stored within last_tx (which is all of them):

```
    def _cleanup(self) -> None:
        while len(self._buffer) > self.maxsize:
            heapq.heappop(self._buffer)

        # This is the problem area:
        while self._buffer and self._buffer[0][0] <= self._last_tx:
            heapq.heappop(self._buffer)
```

now, when push calls the _update_has_item function, we always hit the "not packet_ready" condition:

```
    def _update_has_item(self, printout="") -> None:
        prefilled = self._prefill == 0
        packet_ready = len(self._buffer) > self.prefsize

        # We will always hit this after 65536 due to buffer being emptied in cleanup
        if not prefilled or not packet_ready:
            self._has_item.clear()
            return

        sequential = self._last_tx + 1 == self._buffer[0][0]
        positive_seq = self._last_tx > 0 or self._generation > 0

        # We have the next packet ready
        # OR we havent sent a packet out yet
        # OR the buffer is full
        if (sequential and positive_seq) or not positive_seq or len(self._buffer) >= self.maxsize:
            self._has_item.set() # this is what NEEDS to be called in order for pop to successfully pop
        else:
            self._has_item.clear()
```
The above is exactly why @brianle-riceroll was observing that the 

you can see why @brianle-riceroll observed `ok = self._has_item.wait(timeout)` always giving a bad value once recording started to fail.  This is because after we surpass 65536, push can never call set again due to cleanup wiping the buffer.

The good news is the fix is simple, simply add a similar check to pop that we do to initially increment self._generation in push:

```
    def pop(self, *, timeout=1.0):
        """
        If timeout is a positive number, wait as long as timeout for a packet
        to be ready and return that packet, otherwise return None.
        """

        ok = self._has_item.wait(timeout)
        if not ok:
            print("not ok")
            return None

        if self._prefill > 0:
            return None

        # This function should actually be redundant but i'll leave it for now
        packet = self._pop_if_ready()

        if packet is not None:
            seq = self._get_seq(packet)
            print(f"pop test: {self._last_tx} | {seq}")

            ##################
            # REMOVE
            ##################
            # self._last_tx = self._get_seq(packet)
            ##################


            ##################
            # ADD
            ##################
            # Detect large jump in data sequence caused by self._generation += 1
            # Tried to 32768 is chosen to mirror push function. Any large values up to 65536 should work.
            if seq + 32768 < self._last_tx:
                self._last_tx = self._get_seq(packet) - 65536
            else:
                self._last_tx = self._get_seq(packet)
            ##################

        self._update_has_item("pop")
        return packet
```

The bad news is that HeapJitterBuffer does not change much between this and what is currently in the latest git repo.  This means the issue likely occurs there too.

**Comment by madman37287** [comment]
Thanks! honestly you were 90% of the way there, and your comment pointing to the wait function saved me a ton of time.  I was planning on slapping together a pr, but perhaps a fork containing the PyPI code would be better, especially for people who want a real-time solution.

**Comment by madman37287** [comment]
for the realtime solution: https://github.com/madman37287/discord-ext-voice-recv-realtime

**Comment by madman37287** [comment]
Thanks!  You were 90% of the way there, and you post on the wait function saved me a ton of debugging time.  A PR is currently up for this issue, although it should be noted that if it goes through, it will be part of the slow git repo.  If anyone needs this to run real-time applications, you will want to manually insert this fix into the pip install version.

**Comment by Pl8tinium** [comment]
why cant we revert the changes that made therepo version slow in the first place and publish a new version with the fixes. Was the refactoring that caused it to be slow so substantial that they need to be kept?

**Comment by brianle-riceroll** [comment]
> why cant we revert the changes that made therepo version slow in the first place and publish a new version with the fixes. Was the refactoring that caused it to be slow so substantial that they need to be kept?

The refactors removed the decoder threads, meaning that for all users in a voice call, it will be handled under one thread rather than spinning up a decoder thread for each user. While this may have simplified much of the decoding process, from my testing, the refactors wrote more unnecessary silent packets to the write() method, slowing it down a lot. This isn't ideal for real-time transcription that I am doing. I haven't done enough testing to see whether or not it fixed the issue that was discussed here.

We're not able to revert the changes in this repo (the original maintainer/dev of this repo has been inactive for almost a year). But we can probably fork the commit version f1dbe14be3dc92b666f99df186053be798707807 (The Dec. 10 version, which is also the pip's version)

**Comment by imayhaveborkedit** [comment]
Hi.  About time I check in I think.  Some recent irl changes and other projects has kept me from working on this for a few months.  I was working on fixing some of the issues but never unfortunately never got to finish and push the code.  Now that I have some time again I can get back to fixing these things.  I already knew about the UDP keepalive (and had it mostly finished) and the lock performance issues but I'll review all the stuff from the recent issues and PRs and see what else needs to be fixed.  Thank you all for your interest in this project.

**Comment by madman37287** [comment]
Thanks @imayhaveborkedit !

Also real quick to continue the thread, there's another smaller half issue related to this that I just fixed:

---------------------------------
Description:
---------------------------------
calls to the flush function can be breaking due to resetting the generation variable that gave us so much issues earlier.  Because this is such a stingy variable that requires proper care, i decided to remove it entirely.  Instead of following along with the sequence from 0 to infinity, I instead make use of the reset function to completely reset last_rx and last_tx to 0 once the buffer exceeds 65535.  This also removes the need for the _gen_seq function, since now we only care about packet.sequence.

---------------------------------
Code Fix:
---------------------------------
file: buffer.py

```
    # Left for compatibility purposes
    def _get_seq(self, packet: RTPPacket) -> int:
        return packet.sequence

    def push(self, packet: RTPPacket) -> bool:
        """
        Push a packet into the buffer.  If the packet would make the buffer
        exceed its maxsize, the oldest packet will be dropped.
        """
       
        ################
        # Change
        ################
        seq = packet.sequence
        # if the seq has rolled over, it'll be significantly lower than last rx seq
        if seq + 32768 < self._last_rx:
            with self._lock: # For thread safety, can be moved elsewhere if necessary
                self.reset()
        ################
        ################
```
```
    def pop(self, *, timeout=1.0):
        """
        If timeout is a positive number, wait as long as timeout for a packet
        to be ready and return that packet, otherwise return None.
        """

        ok = self._has_item.wait(timeout)
        if not ok:
            return None

        if self._prefill > 0:
            return None

        # This function should actually be redundant but i'll leave it for now
        packet = self._pop_if_ready()

        if packet is not None:
            ################
            # Change
            ################
            # Completely removed old solution and directly call packet.sequence
            with self._lock: # For thread safety since last tx is also set in reset
                self._last_tx = packet.sequence
            ################
            ################

        self._update_has_item()
        return packet
```

**Comment by imayhaveborkedit** [comment]
I've pushed some commits that should fix two of the issues, the keepalive and the new encryption mode.  Still working on sorting out the other ones.

**Comment by AWAS666** [comment]
> I've pushed some commits that should fix two of the big issues, the keepalive and the new encryption mode. Still working on sorting out the other ones.

Not sure if that is on your radar, but with the new changes it is noticeably slower to receive packages on my end, the old version was far quicker there....

**Comment by brianle-riceroll** [comment]
> > I've pushed some commits that should fix two of the big issues, the keepalive and the new encryption mode. Still working on sorting out the other ones.
> 
> Not sure if that is on your radar, but with the new changes it is noticeably slower to receive packages on my end, the old version was far quicker there....

Yes, he is aware of the issue.

**Comment by imayhaveborkedit** [comment]
Finally I think this is fixed now with bd357957fcfac1f53b526fd6414af73278ecdaf9.  I'll leave this open for a little while just in case I missed something.

**Comment by adbreeker** [comment]
Hi I believe I have similar problem. After few minutes (different each time) of working just fine, my bot is no longer receiving voice packets. It always happens right in the moment when last discord.ext.voice_recv.router: dispatch is logged (on attached picture). Disconnecting and connecting bot again to voice channel fixes it but again only for a few minutes.

![Image](https://github.com/user-attachments/assets/31e6e799-7e8d-4811-8cf2-c6cdad2b9f30)

**Comment by kiley-poole** [comment]
I had the same issue as above. Went through everyone's fixes. The only thing that works is what was just suggested of having the bot emit silent noise the entire time.
